pkg/component/controller/cplb_unix.go

```diff
@@ -0,0 +1,328 @@
+		if !slices.Contains(strAddrs, addr) {
+			if err = k.setLinkIP(addr, linkName, link); err != nil {
+				return fmt.Errorf("failed to add address %s to link %s: %v", addr, linkName, err)
+			}
+		}
+	}
+
+	return nil
+}
+
+func (k *Keepalived) setLinkIP(addr string, linkName string, link netlink.Link) error {
+	ipAddr, _, err := net.ParseCIDR(addr)
+	if err != nil {
+		return fmt.Errorf("failed to parse CIDR %s: %v", addr, err)
+	}
+
+	var mask net.IPMask
+	if ipAddr.To4() != nil {
+		mask = net.CIDRMask(32, 32)
+	} else {
+		mask = net.CIDRMask(128, 128)
+	}
+
+	linkAddr := &netlink.Addr{
+		IPNet: &net.IPNet{
+			IP:   ipAddr,
+			Mask: mask,
+		},
+	}
+
+	k.log.Infof("Adding address %s to link %s", addr, linkName)
+	if err := netlink.AddrAdd(link, linkAddr); err != nil {
+		return fmt.Errorf("failed to add address %s to link %s: %v", addr, linkName, err)
+	}
+	return nil
+}
+
+func (*Keepalived) getLinkAddresses(link netlink.Link) ([]netlink.Addr, []string, error) {
+	linkAddrs, err := netlink.AddrList(link, netlink.FAMILY_ALL)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to list addresses for link %s: %v", link.Attrs().Name, err)
+	}
+
+	strAddrs := make([]string, len(linkAddrs))
+	for i, addr := range linkAddrs {
+		strAddrs[i] = addr.IPNet.String()
+	}
+	return linkAddrs, strAddrs, nil
+}
+
+func (k *Keepalived) generateKeepalivedTemplate() error {
+	f, err := os.OpenFile(k.K0sVars.KeepalivedConfigFile, os.O_CREATE|os.O_WRONLY|os.O_TRUNC, fs.FileMode(0500))
+	if err != nil {
+		return fmt.Errorf("failed to open keepalived config file: %v", err)
+	}
+	defer f.Close()
+
+	template, err := template.New("keepalived").Parse(keepalivedConfigTemplate)
+	if err != nil {
+		return fmt.Errorf("failed to parse keepalived template: %v", err)
+	}
+
+	kc := keepalivedConfig{
+		VRRPInstances: k.Config.VRRPInstances,
+	}
+	if err = template.Execute(f, kc); err != nil {
+		return fmt.Errorf("failed to execute keepalived template: %v", err)
+	}
+
+	// TODO: Do we really need to this every single time?
+	if err = os.Chown(k.K0sVars.KeepalivedConfigFile, k.uid, -1); err != nil {
+		return fmt.Errorf("failed to chown keepalived config file: %v", err)
+	}
+	if err = os.Chmod(k.K0sVars.KeepalivedConfigFile, fs.FileMode(0400)); err != nil {
+		return fmt.Errorf("failed to chmod keepalived config file: %v", err)
+	}
+	return nil
+}
+
+// keepalivedConfig contains all the information required by the
+// KeepalivedConfigTemplate.
+// Right now this struct doesn't make sense right now but we need this for the
+// future virtual_server support.
+type keepalivedConfig struct {
+	VRRPInstances []k0sAPI.VRRPInstance
+}
+
+const dummyLinkName = "dummyvip0"
+const keepalivedConfigTemplate = `
+{{ range .VRRPInstances }}
+vrrp_instance {{ .Name }} {
+    state BACKUP
+    # Make sure the interface is aligned with your server's network interface
+    interface {{ .Interface }}
+    # The virtual router ID must be unique to each VRRP instance that you define
```

Comment:
Hmm, not sure what `VRRP instance` means in this context, a cluster? If it is cluster, maybe use that term here?

Reply:
I think this is a great point but I'd say the right place to explain this is in the docs. I added a new section called `VRRP Instances` in `docs/cplb.md` which should address this. Let me know what you think about it.